### PR TITLE
Make ConfigurationBinder support value types

### LINKED ISSUE #36023: Cannot bind configuration to a value type
### Describe the bug
Not sure if this is actually a bug or if it's intentional, but it's apparently not possible to bind configuration to a value type.

### To Reproduce
1. Using version `3.0.0` of package `Microsoft.Extension.Configuration.Binder`
2. Run this code

```csharp
static void Main()
{
    var configuration = new ConfigurationBuilder()
        .AddInMemoryCollection(new Dictionary<string, string>
        {
            ["X"] = "hello",
            ["Y"] = "world"
        })
        .Build();
        
    var foo = configuration.Get<Foo>();
}

public struct Foo
{
    public string X { get; set; }
    public string Y { get; set; }
}
```
3. The following exception is thrown: **InvalidOperationException**: Cannot create instance of type 'ConsoleApplication1.Program+Foo' because it is missing a public parameterless constructor.

### Expected behavior
The binding should succeed. It works fine if `Foo` is a class. And `Foo` is a value type, so it doesn't *need* a public parameterless constructor.

### Screenshots
N/A

### Additional context
N/A

### PR DESCRIPTION
**Pull request by thomaslevesque**
Fixes #36023

### CODE REVIEW

File: src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs [hunk 1]
```diff
@@ -367,10 +367,13 @@ private static object CreateInstance(Type type)
                 return Array.CreateInstance(typeInfo.GetElementType(), 0);
             }
 
-            bool hasDefaultConstructor = typeInfo.DeclaredConstructors.Any(ctor => ctor.IsPublic && ctor.GetParameters().Length == 0);
-            if (!hasDefaultConstructor)
+            if (!typeInfo.IsValueType)
```
Comment (ericstj) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs:370]: Do we need to disallow internal value types that a caller wouldn't normally have access to?  We probably don't need to be any more restrictive than we are for classes, but I wasn't sure what was allowed for classes.

Comment (ericstj) [reply src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs:370]: Looks like this code already allows creation of an internal type in another assembly if that type had a public constructor, so it looks like this doesn't make that any worse.

**Review by ericstj — APPROVED** [review]
LGTM, thank you for the contribution!

**Review by maryamariyan — APPROVED** [review]

**Pull request merged**